File: tools/plato-spike/clash/validation/ts/adapter.ts
/* Tuple-API adapter over the Plato-generated clash math (plato.g.ts).
 *
 * Exposes the SAME signatures as the hand-written originals
 * (packages/clash/src/math/{vec3,aabb,triangle-intersect}.ts) but every
 * operation is delegated to the generated Vec3 / Box3 classes. Nothing here
 * re-implements the math; it only marshals tuple <-> Vec3 and AABB <-> Box3.
 *
 * Importing plato.g.ts also installs the Number/Boolean prototype intrinsics
 * (Add, Subtract, Max2, FoldMin, ...) that the generated methods rely on. */

import type { Vec3, AABB } from './types.js';
import { Vec3 as PVec3, Box3 as PBox3 } from './plato.g.js';

// ---- marshalling helpers ----
function v(a: Vec3): PVec3 {
  return new PVec3(a[0], a[1], a[2]);
}
function toTuple(p: PVec3): Vec3 {
  return [p.X, p.Y, p.Z];
}
function box(b: AABB): PBox3 {
  return new PBox3(v(b.min), v(b.max));
}
function toAABB(b: PBox3): AABB {
  return { min: toTuple(b.Min), max: toTuple(b.Max) };
}

// ---- vec3 primitives ----
export function sub(a: Vec3, b: Vec3): Vec3 {
  return toTuple(v(a).Sub(v(b)));
}

export function add(a: Vec3, b: Vec3): Vec3 {
  return toTuple(v(a).Add(v(b)));
}

export function scale(a: Vec3, s: number): Vec3 {
  return toTuple(v(a).Scale(s));
}

export function cross(a: Vec3, b: Vec3): Vec3 {
  return toTuple(v(a).Cross(v(b)));
}

export function dot(a: Vec3, b: Vec3): number {
  return v(a).Dot(v(b));
}

export function lenSq(a: Vec3): number {
  return v(a).LenSq();
}

export function distSq(a: Vec3, b: Vec3): number {
  return v(a).DistSq(v(b));
}

export function mid(a: Vec3, b: Vec3): Vec3 {
  return toTuple(v(a).Mid(v(b)));
}

export function centroid(a: Vec3, b: Vec3, c: Vec3): Vec3 {
  return toTuple(v(a).Centroid(v(b), v(c)));
}

// ---- aabb / box primitives ----
export function inflate(b: AABB, m: number): AABB {
  return toAABB(box(b).Inflate(m));
}

export function center(b: AABB): Vec3 {
  return toTuple(box(b).Center());
}

export function intersects(a: AABB, b: AABB): boolean {
  return box(a).Intersects(box(b));
}

export function signedGap(a: AABB, b: AABB): number {
  return box(a).SignedGap(box(b));
}

export function overlapBounds(a: AABB, b: AABB): AABB {
  return toAABB(box(a).OverlapBounds(box(b)));
}

export function boundsOfPoints(a: Vec3, b: Vec3): AABB {
  return toAABB(v(a).BoundsOfPoints(v(b)));
}

export function aabbContains(outer: AABB, inner: AABB): boolean {
  return box(outer).Contains(box(inner));
}

// ---- triangle-triangle SAT ----
export function triTriIntersect(
  a0: Vec3,
  a1: Vec3,
  a2: Vec3,
  b0: Vec3,
  b1: Vec3,
  b2: Vec3,
  eps = 1e-12,
): boolean {
  return v(a0).TriTriIntersectEps(v(a1), v(a2), v(b0), v(b1), v(b2), eps);
}

File: tools/plato-spike/clash/validation/rust/src/adapter.rs
//! Reference-shaped API ([f64; 3] vectors, min/max Aabb struct) implemented
//! ONLY in terms of the generated `plato` module. No math is done here: every
//! function converts the argument shapes and delegates to plato.

use crate::plato;

/// A 3-component vector stored as `[x, y, z]` (reference shape).
pub type Vec3 = [f64; 3];

/// An axis-aligned bounding box with explicit `min`/`max` corners (reference shape).
#[derive(Clone, Copy, PartialEq, Debug)]
pub struct Aabb {
    pub min: [f64; 3],
    pub max: [f64; 3],
}

#[inline]
fn pv(v: Vec3) -> plato::Vec3 {
    plato::Vec3::new(v[0], v[1], v[2])
}

#[inline]
fn vp(v: plato::Vec3) -> Vec3 {
    [v.X, v.Y, v.Z]
}

#[inline]
fn pb(b: &Aabb) -> plato::Box3 {
    plato::Box3::new(pv(b.min), pv(b.max))
}

#[inline]
fn bp(b: plato::Box3) -> Aabb {
    Aabb {
        min: vp(b.Min),
        max: vp(b.Max),
    }
}

// ---- vec3 surface -----------------------------------------------------------

#[inline]
pub fn sub(a: Vec3, b: Vec3) -> Vec3 {
    vp(pv(a).Sub(pv(b)))
}

#[inline]
pub fn add(a: Vec3, b: Vec3) -> Vec3 {
    vp(pv(a).Add(pv(b)))
}

#[inline]
pub fn scale(a: Vec3, s: f64) -> Vec3 {
    vp(pv(a).Scale(s))
}

#[inline]
pub fn cross(a: Vec3, b: Vec3) -> Vec3 {
    vp(pv(a).Cross(pv(b)))
}

#[inline]
pub fn dot(a: Vec3, b: Vec3) -> f64 {
    pv(a).Dot(pv(b))
}

#[inline]
pub fn dist_sq(a: Vec3, b: Vec3) -> f64 {
    pv(a).DistSq(pv(b))
}

#[inline]
pub fn mid(a: Vec3, b: Vec3) -> Vec3 {
    vp(pv(a).Mid(pv(b)))
}

#[inline]
pub fn centroid(a: Vec3, b: Vec3, c: Vec3) -> Vec3 {
    vp(pv(a).Centroid(pv(b), pv(c)))
}

// ---- aabb surface -----------------------------------------------------------

impl Aabb {
    #[inline]
    pub fn new(min: [f64; 3], max: [f64; 3]) -> Self {
        Self { min, max }
    }

    #[inline]
    pub fn inflate(&self, m: f64) -> Aabb {
        bp(pb(self).Inflate(m))
    }

    #[inline]
    pub fn center(&self) -> Vec3 {
        vp(pb(self).Center())
    }

    #[inline]
    pub fn intersects(&self, b: &Aabb) -> bool {
        pb(self).Intersects(pb(b))
    }
}

#[inline]
pub fn signed_gap(a: &Aabb, b: &Aabb) -> f64 {
    pb(a).SignedGap(pb(b))
}

#[inline]
pub fn overlap_bounds(a: &Aabb, b: &Aabb) -> Aabb {
    bp(pb(a).OverlapBounds(pb(b)))
}

#[inline]
pub fn bounds_of_points(a: Vec3, b: Vec3) -> Aabb {
    bp(pv(a).BoundsOfPoints(pv(b)))
}

#[inline]
pub fn aabb_contains(outer: &Aabb, inner: &Aabb) -> bool {
    pb(outer).Contains(pb(inner))
}

// ---- triangle surface --------------------------------------------------------

#[inline]
pub fn tri_tri_intersect(a0: Vec3, a1: Vec3, a2: Vec3, b0: Vec3, b1: Vec3, b2: Vec3) -> bool {
    pv(a0).TriTriIntersect(pv(a1), pv(a2), pv(b0), pv(b1), pv(b2))
}
